Update to client-go 10.0.0

lblackstone opened the pull request:
- Switch build to use Go 1.11 with modules
- Update outdated references to dynamic client functions

Fixes #344

hausdorff commented:
I'm a bit confused on the use of Go modules -- is this required? I'd rather completely use either dep or Go modules if we can, using both is weird to me.

lblackstone commented:
@hausdorff I was having a tough time getting it to build with dep, but it worked well with modules. I can keep trying with dep, but I think life would be easier if we start moving to modules.

lblackstone commented:
@lukehoban I spent several hours yesterday unsuccessfully trying to get this building with dep. The k8s client-go library is notoriously finicky due to the large number of dependencies. I could probably get it working with some more effort, but I would prefer to move to the use of Go modules across the board.

In general, I'd like to move away from building directly on the host to building in a container. Using Docker as the abstraction also allows us to change build tooling in the future without requiring users to care about the toolchain.

_Note: You can still use modules within a GOPATH, but the `GO111MODULE=on` env var must be set (it's default for repos outside of the GOPATH)_

hausdorff commented:
I'm less concerned than @lukehoban is about using dep in some places and Go modules in others. I'm much more concerned with the fact that (1) Kubernetes itself has not transitioned, and (2) there do not seem to be any "serious" production systems using client-go via Go modules. I'm additionally skeptical (based on my experience with kubespy) that upgrading will be easier. Could be wrong -- but I'd personally want to be sure before we move.

If Go modules really work and dep really doesn't, it seems like we could just use those SHAs in the dep file and it should be solved incidentally?

lblackstone commented:
@hausdorff Just chatted with Luke offline about this, and decided to try again with dep using the SHAs from the go.mod file. If I still can't get that working, we'll regroup and decide how heavily to invest in Go modules across the org at this time.

lblackstone commented:
I was able to get the provider building again with dep, but now I'm troubleshooting regressions introduced by the upgrade.

lblackstone commented:
I'm close to having this working now. The client currently isn't working right for resource kinds that don't have a namespace, but I know what the problem is, and am working on squashing bugs.

I'm also investigating a failure on https://github.com/pulumi/pulumi-kubernetes/blob/69d6a815968ab99fa6cfb409176df1fad031e793/tests/integration/get/step1/index.ts#L41-L45 where the preview fails due to the CRD not being registered yet. I'm still unclear if this is a bug in the test, or if I inadvertently changed the semantics there.

hausdorff commented:
@lblackstone I think I mentioned this offline, but just to follow up on the integration test before, we could put the `.get` in a `step2/` folder -- no one would ever try to do something like that in real life, but it would give us code coverage for that path.

That said, I'm somewhat confused how this was passing before. :)

lblackstone commented:
I squashed and rebased on master in the last force-push.

lblackstone commented:
Rebased on master and resolved merge conflict.

hausdorff commented on pkg/await/await.go:
```diff
@@ -229,51 +253,49 @@ func Update(
-	currentOutputs, err := clientForResource.Patch(currentSubmitted.GetName(), patchType, patch)
+	// Issue patch request.
+	// NOTE: We can use the same client because if the `kind` changes, this will cause
+	//		 a replace (i.e., destroy and create).
```
Weird tab here -- I've seen this before in your PRs so my guess is that your editor (JetBrains?) does "stuff", lol. Would you mind double checking that there aren't other issues like this?

metral replied:
Can you use `gofmt` in your editor? With the general styling & format comments in this PR, I think we should setup go formatting, linting, etc in a git commit hook. Thoughts?

lblackstone replied:
I ran this through `goimports` already which should also run `gofmt`. I can't remember if the build toolchain includes that, but I agree that we should set it up if not.

hausdorff commented on pkg/await/awaiters.go:
```diff
@@ -567,7 +567,7 @@ func untilCoreV1ServiceAccountInitialized(c createAwaitConfig) error {
 	//
 
-	specSecrets, _ := openapi.Pluck(c.currentInputs.Object, "secrets")
+	specSecrets, _ := openapi.Pluck(c.currentOutputs.Object, "secrets")
```
Why the change from inputs to outputs?

metral replied:
I had the same question - but just assumed that it was me not having enough context as a novice

hausdorff commented on pkg/clients/clients.go:
```diff
@@ -0,0 +1,390 @@
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(discoCacheClient)
+
+	// Create dynamic resource client
+	client, err := dynamic.NewForConfig(clientConfig)
```
Is this client still using the REST pool?

lblackstone replied:
It's implicit now in the dynamic client.

hausdorff commented on pkg/await/core_service.go:
```diff
@@ -253,11 +240,11 @@ func (sia *serviceInitAwaiter) await(
 
 func (sia *serviceInitAwaiter) processServiceEvent(event watch.Event) {
-	inputServiceName := sia.config.currentInputs.GetName()
+	inputServiceName := sia.config.currentOutputs.GetName()
```
Another change from inputs to outputs here, this one seems to be harmless. Just wondering if it was on purpose.

lblackstone replied:
I'll revert and see. Can't remember why at this point.

hausdorff commented on pkg/await/retry.go:
```diff
@@ -41,7 +56,7 @@ func (r *retrier) Do() error {
 		err = r.try(r.tries)
 		r.tries++
-		if errors.IsNotFound(err) {
+		if errors.IsNotFound(err) || meta.IsNoMatchError(err) {
```
I assume this is for the CRDs. This kind of scares me. Is client-go changing the error API? And if so, how do we know what else to check?

lblackstone replied:
Yes, this is required for the RESTMapper, which will return an unknown Kind error if the relevant CRD hasn't been registered in the cache yet.

hausdorff commented on pkg/provider/provider.go:
```diff
@@ -533,7 +539,8 @@ func (k *kubeProvider) Read(ctx context.Context, req *pulumirpc.ReadRequest) (*p
 	}
 
-	id := client.FqObjName(liveObj)
+	// TODO(lblackstone): not sure why this is needed
```
Me neither.

lblackstone replied:
I figured I'd go back over the minor TODOs in a separate PR.

metral commented on pkg/await/apps_deployment.go:
```diff
@@ -558,13 +556,13 @@ func (dia *deploymentInitAwaiter) aggregatePodErrors() ([]string, []string) {
 	}
 
-	scheduleErrors := []string{}
+	scheduleErrors := make([]string, 0)
```
I'm a struct literal user myself, but they're all equivalent - `make` with a size of 0 is more explicit for the reader.

All of these are valid and identical:
- `[]string{}`
- `make([]string, 0)`
- `make([]string)`

lblackstone replied:
Again, this was due to my IDE complaining ("Empty slice declaration via literal").
I believe the rationale is that the `make` syntax is more obviously allocating memory for the slice, which would not be the case for, e.g., `var scheduleErrors []string`. As Mike points out, the updated syntax is semantically equivalent.

hausdorff replied:
Ok, let's just standardize on `make(whatever, 0)` then.

lblackstone commented on Gopkg.toml:
```diff
@@ -8,53 +8,59 @@ required = ["k8s.io/kubernetes/pkg/kubectl/cmd/util"]
 [[override]]
-  name = "k8s.io/apiserver"
-  version = "kubernetes-1.10.2"
+  name = "sigs.k8s.io/kustomize"
```
I think it's part of kubectl now. Not sure why it got pulled in.

lblackstone commented on examples/provider/index.ts:
```diff
@@ -20,7 +19,7 @@ const nginxcontainer = new k8s.core.v1.Pod("nginx", {
             image: "nginx:1.7.9",
             name: "nginx",
-            ports: [{ containerPort: 80 }],
+            ports: [{containerPort: 80}],
```
I guess my IDE auto-formatted when I saved changes.

lblackstone commented on pkg/clients/clients.go:
```diff
@@ -0,0 +1,390 @@
+	ServiceAccount                 Kind = "ServiceAccount"
+	StatefulSet                    Kind = "StatefulSet"
+	StorageClass                   Kind = "StorageClass"
+	ValidatingWebhookConfiguration Kind = "ValidatingWebhookConfiguration"
```
Right. I wasn't sure if there was a cleaner way to handle this. Looks like it might be possible to introspect from the spec?

lblackstone commented on pkg/clients/clients.go:
```diff
@@ -0,0 +1,390 @@
+	switch k {
+	case APIService, CertificateSigningRequest, ClusterRole, ClusterRoleBinding, CustomResourceDefinition,
+		MutatingWebhookConfiguration, Namespace, PersistentVolume, PodSecurityPolicy, PriorityClass,
+		StorageClass, ValidatingWebhookConfiguration:
```
Yeah, wasn't sure how to do that, but that's certainly preferable.

hausdorff replied:
I'm not sure there's an easy way, but one way is to check the `.paths` field of the API and see if the create operation has a variant with `/{namespace}/` in it.

lblackstone replied:
I'll add that as a TODO, and circle back on it.

lblackstone replied:
Rewrote this with introspection from the discovery client.

lblackstone commented on pkg/clients/clients.go:
```diff
@@ -0,0 +1,390 @@
+	// Cache the discovery information (OpenAPI schema, etc.) so we don't have to retrieve it for
+	// every request.
+	discoCacheClient := cacheddiscovery.NewMemCacheClient(disco)
+	discoCacheClient.Invalidate()
```
No, it seems not. The call will fail due to a cache miss without that.

lblackstone commented on pkg/await/version.go:
```diff
@@ -23,11 +23,10 @@ import (
 
 // Format v0.0.0(-master+$Format:%h$)
-var gitVersionRe = regexp.MustCompile("v([0-9])+.([0-9])+.[0-9]+.*")
+var gitVersionRe = regexp.MustCompile(`v([0-9])+.([0-9])+.[0-9]+.*`)
```
Backticks are literal strings, which are much nicer for regexs. Doesn't matter on this one, but it generally avoids the need for escaping certain characters. No downside, so it's a good habit.

lblackstone commented on pkg/await/await.go:
```diff
@@ -319,10 +337,16 @@ func Deletion(
+		if v, err := parseVersion(sv); err == nil {
+			version = v
+		} else {
+			version = DefaultVersion()
```
Updated defaultVersion. ServerVersion comes from the vendored discovery lib.

hausdorff replied:
Isn't it defined in `pkg/await/version.go`?

lblackstone replied:
Whoops, I assumed you meant the `ServerVersion()` call on 342. Will update.

lblackstone commented on pkg/await/awaiters.go:
```diff
@@ -276,7 +263,7 @@ func untilAppsDeploymentDeleted(
 
 		return watcher.RetryableError(
-			fmt.Errorf("Deployment '%s' still exists (%d / %d replicas exist)", name,
+			fmt.Errorf("deployment %q still exists (%d / %d replicas exist)", name,
```
https://github.com/golang/go/wiki/CodeReviewComments#error-strings :)
Most of the minor changes were to fix warnings in my IDE.

lblackstone commented:
So the discovery interface is pretty cool for introspecting on the k8s API, but it turns out to be really flaky when you're slamming it pretty hard (e.g., the istio test). I added a default look up table back for resource namespacing that will fall back to server-side discovery, and also added retry logic around discovery calls to make it more robust. I was able to spin the istio example up a couple times without error, so I think it's good to go now. 🤞

lblackstone commented:
Fixed the istio test issues and seem to have introduced a regression on the Read, probably related to the oldInputs/currentInputs comments in the review. Working on that now.

hausdorff commented on pkg/clients/clients.go:
```diff
@@ -375,6 +141,55 @@ func (dcs *DynamicClientSet) ResourceClientForObject(obj *unstructured.Unstructu
 }
 
+func (dcs *DynamicClientSet) gvkForKind(kind Kind) (gvk schema.GroupVersionKind, err error) {
+	resources, err := dcs.DiscoveryClientCached.ServerPreferredResources()
```
Just so I'm clear about this: we're using `ServerPreferredResources` here because the API server, when used with this client, transparently converts various API versions into the "preferred" one, right? Also which versions of k8s does this work with.

lblackstone replied:
Yes, that's my understanding. I've actually modified this logic somewhat so that it uses the provided GV by default. I think it should work for any k8s.

hausdorff commented on pkg/clients/clients.go:
```diff
@@ -375,6 +141,55 @@ func (dcs *DynamicClientSet) ResourceClientForObject(obj *unstructured.Unstructu
+	return true
+}
+
+func parseGVString(gv string) schema.GroupVersion {
```
I feel like there should be an official parsing function for this?

lblackstone replied:
Ah, you're right. Will update.

hausdorff commented on pkg/clients/clients.go:
```diff
@@ -161,25 +167,52 @@ func (dcs *DynamicClientSet) gvkForKind(kind Kind) (gvk schema.GroupVersionKind,
+	switch Kind(gvk.Kind) {
+	case APIService, CertificateSigningRequest, ClusterRole, ClusterRoleBinding, CustomResourceDefinition,
+		MutatingWebhookConfiguration, Namespace, PersistentVolume, PodSecurityPolicy, PriorityClass,
+		StorageClass, ValidatingWebhookConfiguration:
```
lol I don't like this but I also want to not block the review. Can we add a comment explaining why we have this lookup table? I understand this to be because the API server can be flaky, which is ???? scary?

lblackstone replied:
I think the flakiness comes from the discovery client. It requires the caller to handle retries when `Invalidate()` is called, and won't return resources while the cache is refreshing. I'll add a comment and TODO to see if there's a cleaner way to handle that.

hausdorff commented on pkg/clients/clients.go:
```diff
@@ -161,25 +167,52 @@ func (dcs *DynamicClientSet) gvkForKind(kind Kind) (gvk schema.GroupVersionKind,
+
+func (dcs *DynamicClientSet) getServerResources() ([]*v1.APIResourceList, error) {
+	var resources []*v1.APIResourceList
+	err := retry.SleepingRetry(
```
Combining multiple exponential backoff conditions might cause some of our timeouts to fail more frequently. We've tried to put the retry logic in one place to make it really predictable. I'm fine shipping this for now, but highly recommend we come back later and audit the retry path and make sure it actually _is_ predictable.

hausdorff commented on pkg/clients/clients.go:
```diff
@@ -191,30 +196,38 @@ func (dcs *DynamicClientSet) namespaced(gvk schema.GroupVersionKind) (bool, erro
 }
 
+func isServerCacheError(err error) bool {
+	if err == cacheddiscovery.ErrCacheEmpty || err == cacheddiscovery.ErrCacheNotFound {
```
Can we follow up to make sure we're catching any other changes to the error API? This changing of the error semantics scares me a lot.

hausdorff commented:
Overall we're getting close. Just a couple more discussion points.

lblackstone commented:
Last failure was the python flake. Addressed remaining feedback and this should hopefully be good to go

lblackstone commented:
Think this is ready to merge now!
